packages/datadog-shimmer/src/shimmer.js

```diff
@@ -0,0 +1,137 @@
+'use strict'
+
+// Use a weak map to avoid polluting the wrapped function/method.
+const unwrappers = new WeakMap()
+
+function copyProperties (original, wrapped) {
+  Object.setPrototypeOf(wrapped, original)
+
+  const props = Object.getOwnPropertyDescriptors(original)
+  const keys = Reflect.ownKeys(props)
+
+  for (const key of keys) {
+    Object.defineProperty(wrapped, key, props[key])
+  }
+}
+
+function wrapFn (original, delegate) {
+  assertFunction(delegate)
```

Comment:
To future-proof these a bit, can we have them return the value that's been checked?

e.g.

```js
delegate = assertFunction(delegate)
```

We'll need this later on to do type refinements (i.e. "narrowing").

Reply:
Not sure I understand what you mean. If we'll need this later, should we just do it later?

Reply:
Yeah I guess it's an easy enough change to do later. Nevermind.